Non native speakers 165

stpapp opened the pull request:
This PR adds the enhancement when a patient is a non-native English speaker, that patient will have a foreign place of birth, #165 

How it works:
A reference file of ethnicities to cities should be created.  Format is:
{"ethnicity1": ["city1", "city2", ... "cityN"],
"ethnicity2": ["city1", "city2", ... "cityN"],
...
"ethnicityN": ["city1", "city2", ... "cityN"]}
File location can be declared in the synthea.properties file with the "generate.geography.foreign.birthplace.default_file" flag.

During the LifecycleModule, when a non-english first language is found instead of calling Location.randomCityName() we instead call Location.randomCityNameByEthnicity() which references the previously mentioned json file containing maps from ethnicities to cities.  Once the list of cities is accessed (based on the patient's ethnicity) a city is chosen based on the patient's assigned random.  If for any reason a city cannot be chosen (no cities for an ethnicity, ethnicity not contained in the file, etc) the method falls back to choosing a city from Location.randomCityName()

Caveats:
- I did very little city research, simply grabbing a list of the 5 most populous cities in each "region" and using those locations in the src/main/resources/foreign_birthplace.json file included in this PR.  - Additionally the Location.randomCityNameByEthnicity() method does not do any selection based on population size like Location.randomCityName() does, mainly because I don't have access (or the time unfortunately) to put together an exhaustive list of world cities (sorry!).
- Currently it's possible in regions that have multiple main languages for a patient to have a birth city where the language isn't the official language.  For example it would be possible for a person with a south_american ethnicity and a first_language of Spanish, but a birth city of Rio de Janerio (where the Official language is Portuguese).  While this is certainly a possibility, it seems more likely that patient would have a different birth city.  Future enhancements could target this area of functionality.


Commits:
- Adds multiple json files to support a limited foreign birthplace
- Updates the Lifecycle.birth() to pick a birth location based on the Person's defined first language
- Updates Location to be able to pull in a json file for foreign birthplaces
- Adds a method to get a random city name by ethnicity

jawalonoski commented:
This is super cool.

jawalonoski commented:
OK, so I tested this by generating 100 people in Texas.

I got an Asian man who speaks Chinese and was born in Chongqing. Great!

Except, it shows up in FHIR as Chongqing, Texas, US. So, you probably could add attributes for birthplace state/country.

Also, his name is Everett935 Streich926, which makes no sense, but one step at a time...

Example:
```json
{
        "resourceType": "Patient",
        "id": "2d015b7e-6ecc-45b7-819d-de630a555264",
        "text": {
          "status": "generated",
          "div": "<div xmlns=\"http://www.w3.org/1999/xhtml\">Generated by <a href=\"https://github.com/synthetichealth/synthea\">Synthea</a>.Version identifier: v2.2.0-49-g05bf356c\n .   Person seed: 6360867500992721840  Population seed: 1544215612361</div>"
        },
        "extension": [
          {
            "url": "http://hl7.org/fhir/us/core/StructureDefinition/us-core-race",
            "extension": [
              {
                "url": "ombCategory",
                "valueCoding": {
                  "system": "urn:oid:2.16.840.1.113883.6.238",
                  "code": "2028-9",
                  "display": "Asian"
                }
              },
              {
                "url": "text",
                "valueString": "Asian"
              }
            ]
          },
          {
            "url": "http://hl7.org/fhir/us/core/StructureDefinition/us-core-ethnicity",
            "extension": [
              {
                "url": "ombCategory",
                "valueCoding": {
                  "system": "urn:oid:2.16.840.1.113883.6.238",
                  "code": "2186-5",
                  "display": "Not Hispanic or Latino"
                }
              },
              {
                "url": "text",
                "valueString": "Not Hispanic or Latino"
              }
            ]
          },
          {
            "url": "http://hl7.org/fhir/StructureDefinition/patient-mothersMaidenName",
            "valueString": "Pura348 Hickle134"
          },
          {
            "url": "http://hl7.org/fhir/us/core/StructureDefinition/us-core-birthsex",
            "valueCode": "M"
          },
          {
            "url": "http://hl7.org/fhir/StructureDefinition/birthPlace",
            "valueAddress": {
              "city": "Chongqing",
              "state": "Texas",
              "country": "US"
            }
          },
          {
            "url": "http://synthetichealth.github.io/synthea/disability-adjusted-life-years",
            "valueDecimal": 0.005735004927900484
          },
          {
            "url": "http://synthetichealth.github.io/synthea/quality-adjusted-life-years",
            "valueDecimal": 25.9942649950721
          }
        ],
        "identifier": [
          {
            "system": "https://github.com/synthetichealth/synthea",
            "value": "57e3c3e4-8a64-45a2-b183-22ec54b10044"
          },
          {
            "type": {
              "coding": [
                {
                  "system": "http://hl7.org/fhir/v2/0203",
                  "code": "MR",
                  "display": "Medical Record Number"
                }
              ],
              "text": "Medical Record Number"
            },
            "system": "http://hospital.smarthealthit.org",
            "value": "57e3c3e4-8a64-45a2-b183-22ec54b10044"
          },
          {
            "type": {
              "coding": [
                {
                  "system": "http://hl7.org/fhir/identifier-type",
                  "code": "SB",
                  "display": "Social Security Number"
                }
              ],
              "text": "Social Security Number"
            },
            "system": "http://hl7.org/fhir/sid/us-ssn",
            "value": "999-22-6633"
          },
          {
            "type": {
              "coding": [
                {
                  "system": "http://hl7.org/fhir/v2/0203",
                  "code": "DL",
                  "display": "Driver's License"
                }
              ],
              "text": "Driver's License"
            },
            "system": "urn:oid:2.16.840.1.113883.4.3.25",
            "value": "S99927651"
          },
          {
            "type": {
              "coding": [
                {
                  "system": "http://hl7.org/fhir/v2/0203",
                  "code": "PPN",
                  "display": "Passport Number"
                }
              ],
              "text": "Passport Number"
            },
            "system": "http://standardhealthrecord.org/fhir/StructureDefinition/passportNumber",
            "value": "X87455154X"
          }
        ],
        "name": [
          {
            "use": "official",
            "family": "Streich926",
            "given": [
              "Everett935"
            ],
            "prefix": [
              "Mr."
            ]
          }
        ],
        "telecom": [
          {
            "system": "phone",
            "value": "555-347-5789",
            "use": "home"
          }
        ],
        "gender": "male",
        "birthDate": "1991-11-23",
        "address": [
          {
            "extension": [
              {
                "url": "http://hl7.org/fhir/StructureDefinition/geolocation",
                "extension": [
                  {
                    "url": "latitude",
                    "valueDecimal": 30.316017
                  },
                  {
                    "url": "longitude",
                    "valueDecimal": -95.469993
                  }
                ]
              }
            ],
            "line": [
              "249 Boehm Landing Apt 50"
            ],
            "city": "Conroe",
            "state": "Texas",
            "postalCode": "77301",
            "country": "US"
          }
        ],
        "maritalStatus": {
          "coding": [
            {
              "system": "http://hl7.org/fhir/v3/MaritalStatus",
              "code": "S",
              "display": "S"
            }
          ],
          "text": "S"
        },
        "multipleBirthBoolean": false,
        "communication": [
          {
            "language": {
              "coding": [
                {
                  "system": "urn:ietf:bcp:47",
                  "code": "zh",
                  "display": "Chinese"
                }
              ],
              "text": "Chinese"
            }
          }
        ]
      }
```

stpapp commented:
Ok, made some updates which resolve the "Chongqing, Texas, US" issue, additionally updated the format of the foreign_birthplace.json to read as follows:
{"ethnicity1": ["city1,state1,country1", "city2,state2,country2", ... "cityN,stateN,countryN"],
"ethnicity2": ["city1,state1,country1", "city2,state2,country2", ... "cityN,stateN,countryN"],
...
"ethnicityN": ["city1,state1,country1", "city2,state2,country2", ... "cityN,stateN,countryN"]}

So now when I run 100 users, I see output like:

Where I have an Asian man, who speaks Chinese and was born in Shanghai
{
            "url": "http://hl7.org/fhir/StructureDefinition/birthPlace",
            "valueAddress": {
              "city": "Shanghai",
              "state": "Shanghai Municipality",
              "country": "CN"
            }
          }

Full Patient section:
{
      "fullUrl": "urn:uuid:6d35c927-ffa2-496b-b8e5-ba26674defff",
      "resource": {
        "resourceType": "Patient",
        "id": "6d35c927-ffa2-496b-b8e5-ba26674defff",
        "meta": {
          "profile": [
            "http://standardhealthrecord.org/fhir/StructureDefinition/shr-entity-Patient"
          ]
        },
        "text": {
          "status": "generated",
          "div": "<div xmlns=\"http://www.w3.org/1999/xhtml\">Generated by <a href=\"https://github.com/synthetichealth/synthea\">Synthea</a>.Version identifier: v2.2.0-40-gf715b380\n .   Person seed: -5429519707410353741  Population seed: 1544293651252</div>"
        },
        "extension": [
          {
            "url": "http://hl7.org/fhir/us/core/StructureDefinition/us-core-race",
            "extension": [
              {
                "url": "ombCategory",
                "valueCoding": {
                  "system": "urn:oid:2.16.840.1.113883.6.238",
                  "code": "2028-9",
                  "display": "Asian"
                }
              },
              {
                "url": "text",
                "valueString": "Asian"
              }
            ]
          },
          {
            "url": "http://hl7.org/fhir/us/core/StructureDefinition/us-core-ethnicity",
            "extension": [
              {
                "url": "ombCategory",
                "valueCoding": {
                  "system": "urn:oid:2.16.840.1.113883.6.238",
                  "code": "2186-5",
                  "display": "Not Hispanic or Latino"
                }
              },
              {
                "url": "text",
                "valueString": "Not Hispanic or Latino"
              }
            ]
          },
          {
            "url": "http://hl7.org/fhir/StructureDefinition/patient-mothersMaidenName",
            "valueString": "Karleen110 Reinger292"
          },
          {
            "url": "http://hl7.org/fhir/us/core/StructureDefinition/us-core-birthsex",
            "valueCode": "F"
          },
          {
            "url": "http://hl7.org/fhir/StructureDefinition/birthPlace",
            "valueAddress": {
              "city": "Shanghai",
              "state": "Shanghai Municipality",
              "country": "CN"
            }
          },
          {
            "url": "http://standardhealthrecord.org/fhir/StructureDefinition/shr-actor-FictionalPerson-extension",
            "valueBoolean": true
          },
          {
            "url": "http://standardhealthrecord.org/fhir/StructureDefinition/shr-entity-FathersName-extension",
            "valueHumanName": {
              "text": "Jewel43 Rippin620"
            }
          },
          {
            "url": "http://standardhealthrecord.org/fhir/StructureDefinition/shr-demographics-SocialSecurityNumber-extension",
            "valueString": "999-68-4997"
          },
          {
            "url": "http://standardhealthrecord.org/fhir/StructureDefinition/shr-entity-Person-extension",
            "valueReference": {
              "reference": "urn:uuid:a71af3e7-c012-4d66-b6d2-321768471707"
            }
          },
          {
            "url": "http://synthetichealth.github.io/synthea/disability-adjusted-life-years",
            "valueDecimal": 0.005735004927900484
          },
          {
            "url": "http://synthetichealth.github.io/synthea/quality-adjusted-life-years",
            "valueDecimal": 2.9942649950720996
          }
        ],
        "identifier": [
          {
            "system": "https://github.com/synthetichealth/synthea",
            "value": "cb258a5c-3df5-4220-9260-8586bc639571"
          },
          {
            "type": {
              "coding": [
                {
                  "system": "http://hl7.org/fhir/v2/0203",
                  "code": "MR",
                  "display": "Medical Record Number"
                }
              ],
              "text": "Medical Record Number"
            },
            "system": "http://hospital.smarthealthit.org",
            "value": "cb258a5c-3df5-4220-9260-8586bc639571"
          },
          {
            "type": {
              "coding": [
                {
                  "system": "http://hl7.org/fhir/identifier-type",
                  "code": "SB",
                  "display": "Social Security Number"
                }
              ],
              "text": "Social Security Number"
            },
            "system": "http://hl7.org/fhir/sid/us-ssn",
            "value": "999-68-4997"
          }
        ],
        "name": [
          {
            "use": "official",
            "family": "Rippin620",
            "given": [
              "Ferne426"
            ]
          }
        ],
        "telecom": [
          {
            "system": "phone",
            "value": "555-562-5326",
            "use": "home"
          }
        ],
        "gender": "female",
        "birthDate": "2014-11-01",
        "address": [
          {
            "extension": [
              {
                "url": "http://hl7.org/fhir/StructureDefinition/geolocation",
                "extension": [
                  {
                    "url": "latitude",
                    "valueDecimal": 33.107224
                  },
                  {
                    "url": "longitude",
                    "valueDecimal": -96.674676
                  }
                ]
              }
            ],
            "line": [
              "554 Schmeler Passage"
            ],
            "city": "Allen",
            "state": "Texas",
            "postalCode": "75002",
            "country": "US"
          }
        ],
        "maritalStatus": {
          "coding": [
            {
              "system": "http://hl7.org/fhir/v3/MaritalStatus",
              "code": "S",
              "display": "Never Married"
            }
          ],
          "text": "Never Married"
        },
        "multipleBirthBoolean": false,
        "communication": [
          {
            "language": {
              "coding": [
                {
                  "system": "urn:ietf:bcp:47",
                  "code": "zh",
                  "display": "Chinese"
                }
              ],
              "text": "Chinese"
            }
          }
        ]
      },
      "request": {
        "method": "POST",
        "url": "Patient"
      }
    }

A note about added attributes:
I added Person.BIRTH_CITY, Person.BIRTH_STATE and Person.BIRTH_COUNTRY attributes to carry the additional details carried in the foreign_birthplace.json file, and updated the various exporters to use those details in place of the existing state and country vars they were using.  However in the case of CSV exports I kept the output as is, simply outputting a concatenated string of city, state, country under the BIRTHPLACE header because I worried about breaking any existing "contracts" that were based on the previously published format.

I'm not sure if that's how you'd like to proceed here and I'm happy to update the csv output to carry city, state, country if you feel that's the correct path forward, just let me know what you think.  

Updating the csv output to city, state, country would simplify the implementation of the Location.randomBirthPlace() and Location.randomBirthplaceByEthnicity() methods.  

Re: Names matching ethnicities, what do you think about filing that as a separate issue?  It does look fairly straightforward to fix, but might take a bit of time to compile a list of proper names based on ethnicity.

Thanks for the feedback and the pointers around adding attributes, appreciate the guidance.

jawalonoski commented:
I'm not terribly worried about breaking existing contracts with the CSV format, as long as the data dictionary in the Wiki is updated appropriately. That being said, that could be a separate pull request if anyone wants this additional data.

jawalonoski commented:
Thank you for the contribution!
